File: circular_maze.java
// circular_maze.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class circular_maze {
    private int rings, sectors;
    private String algo;
    private long seed;
    private boolean solve;
    private Cell[][] maze;
    private List<int[]> path;
    private Random rand;

    static class Cell {
        boolean v, h;
        Cell(boolean v, boolean h) { this.v = v; this.h = h; }
    }

    public circular_maze(int rings, int sectors, String algo, long seed, boolean solve) {
        this.rings = rings;
        this.sectors = sectors;
        this.algo = algo;
        this.seed = seed;
        this.solve = solve;
        this.rand = seed != 0 ? new Random(seed) : new Random();
    }

    private List<int[]> getNeighbors(int ring, int sector) {
        List<int[]> neighbors = new ArrayList<>();
        if (ring > 0) neighbors.add(new int[]{ring-1, sector, 0});
        if (ring < rings-1) neighbors.add(new int[]{ring+1, sector, 1});
        neighbors.add(new int[]{ring, (sector-1+sectors)%sectors, 2});
        neighbors.add(new int[]{ring, (sector+1)%sectors, 3});
        return neighbors;
    }

    private void generateDFS() {
        maze = new Cell[rings][sectors];
        for (int r = 0; r < rings; r++) {
            for (int s = 0; s < sectors; s++) {
                maze[r][s] = new Cell(true, true);
            }
        }

        Stack<int[]> stack = new Stack<>();
        stack.push(new int[]{0, 0});
        Set<String> visited = new HashSet<>();
        visited.add("0,0");

        while (!stack.isEmpty()) {
            int[] pos = stack.peek();
            int ring = pos[0], sector = pos[1];
            List<int[]> neighbors = getNeighbors(ring, sector);
            Collections.shuffle(neighbors, rand);

            boolean found = false;
            for (int[] n : neighbors) {
                int nr = n[0], ns = n[1], dir = n[2];
                String key = nr + "," + ns;
                if (visited.contains(key)) continue;
                visited.add(key);
                if (dir == 0) maze[nr][ns].h = false;
                else if (dir == 1) maze[ring][sector].h = false;
                else if (dir == 2) maze[ring][(sector-1+sectors)%sectors].v = false;
                else maze[ring][sector].v = false;
                stack.push(new int[]{nr, ns});
                found = true;
                break;
            }
            if (!found) stack.pop();
        }
    }

    public void generate() {
        generateDFS();
        if (solve) path = solveMaze();
    }

    private List<int[]> solveMaze() {
        Stack<List<int[]>> stack = new Stack<>();
        List<int[]> start = new ArrayList<>();
        start.add(new int[]{0, 0});
        stack.push(start);
        Set<String> visited = new HashSet<>();
        visited.add("0,0");

        while (!stack.isEmpty()) {
            List<int[]> path = stack.pop();
            int[] last = path.get(path.size()-1);
            int ring = last[0], sector = last[1];
            if (ring == rings-1) return path;

            for (int[] n : getNeighbors(ring, sector)) {
                int nr = n[0], ns = n[1], dir = n[2];
                String key = nr + "," + ns;
                if (visited.contains(key)) continue;
                boolean hasWall = false;
                if (dir == 0) hasWall = maze[nr][ns].h;
                else if (dir == 1) hasWall = maze[ring][sector].h;
                else if (dir == 2) hasWall = maze[ring][(sector-1+sectors)%sectors].v;
                else hasWall = maze[ring][sector].v;
                if (!hasWall) {
                    visited.add(key);
                    List<int[]> newPath = new ArrayList<>(path);
                    newPath.add(new int[]{nr, ns});
                    stack.push(newPath);
                }
            }
        }
        return null;
    }

    public void printASCII() {
        System.out.println("\u001B[36m\nКарта лабиринта (вид сверху):\u001B[0m");
        System.out.printf("\u001B[33m  Колец: %d, Секторов: %d\u001B[0m\n", rings, sectors);
        for (int ring = 0; ring < rings; ring++) {
            System.out.print("R" + ring + ": ");
            for (int sector = 0; sector < sectors; sector++) {
                System.out.print((maze[ring][sector].v ? "|" : " ") + (maze[ring][sector].h ? "─" : " "));
            }
            System.out.println();
        }
    }

    public void saveJSON(String filename) throws IOException {
        // Упрощённо
        System.out.println("\u001B[32m💾 Сохранено JSON: " + filename + "\u001B[0m");
    }

    public static void main(String[] args) throws Exception {
        int rings = 8, sectors = 12;
        String algo = "dfs";
        long seed = 0;
        boolean solve = false;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--rings")) rings = Integer.parseInt(args[++i]);
            else if (args[i].equals("--sectors")) sectors = Integer.parseInt(args[++i]);
            else if (args[i].equals("--algo")) algo = args[++i];
            else if (args[i].equals("--seed")) seed = Long.parseLong(args[++i]);
            else if (args[i].equals("--solve")) solve = true;
        }

        System.out.println("\u001B[36m⭕ Circular Maze Generator (Java)\u001B[0m");
        System.out.printf("📐 Параметры: %d колец, %d секторов, алгоритм: %s\n", rings, sectors, algo);

        circular_maze gen = new circular_maze(rings, sectors, algo, seed, solve);
        gen.generate();
        gen.printASCII();
        gen.saveJSON("maze.json");
    }
}

File: circular_maze.cs
// circular_maze.cs — C# версия

using System;
using System.Collections.Generic;

class Cell {
    public bool V { get; set; }
    public bool H { get; set; }
    public Cell(bool v, bool h) { V = v; H = h; }
}

class CircularMaze {
    private int rings, sectors;
    private string algo;
    private int seed;
    private bool solve;
    private Cell[][] maze;
    private List<(int, int)> path;
    private Random rand;

    public CircularMaze(int rings, int sectors, string algo, int seed, bool solve) {
        this.rings = rings;
        this.sectors = sectors;
        this.algo = algo;
        this.seed = seed;
        this.solve = solve;
        this.rand = seed != 0 ? new Random(seed) : new Random();
    }

    private List<(int, int, int)> GetNeighbors(int ring, int sector) {
        var neighbors = new List<(int, int, int)>();
        if (ring > 0) neighbors.Add((ring-1, sector, 0));
        if (ring < rings-1) neighbors.Add((ring+1, sector, 1));
        neighbors.Add((ring, (sector-1+sectors)%sectors, 2));
        neighbors.Add((ring, (sector+1)%sectors, 3));
        return neighbors;
    }

    private void GenerateDFS() {
        maze = new Cell[rings][];
        for (int r = 0; r < rings; r++) {
            maze[r] = new Cell[sectors];
            for (int s = 0; s < sectors; s++) maze[r][s] = new Cell(true, true);
        }

        var stack = new Stack<(int, int)>();
        stack.Push((0, 0));
        var visited = new HashSet<(int, int)> { (0, 0) };

        while (stack.Count > 0) {
            var (ring, sector) = stack.Peek();
            var neighbors = GetNeighbors(ring, sector);
            // Shuffle
            for (int i = neighbors.Count-1; i > 0; i--) {
                int j = rand.Next(i+1);
                var temp = neighbors[i];
                neighbors[i] = neighbors[j];
                neighbors[j] = temp;
            }

            bool found = false;
            foreach (var (nr, ns, dir) in neighbors) {
                if (visited.Contains((nr, ns))) continue;
                visited.Add((nr, ns));
                if (dir == 0) maze[nr][ns].H = false;
                else if (dir == 1) maze[ring][sector].H = false;
                else if (dir == 2) maze[ring][(sector-1+sectors)%sectors].V = false;
                else maze[ring][sector].V = false;
                stack.Push((nr, ns));
                found = true;
                break;
            }
            if (!found) stack.Pop();
        }
    }

    public void Generate() {
        GenerateDFS();
        if (solve) path = SolveMaze();
    }

    private List<(int, int)> SolveMaze() {
        var stack = new Stack<List<(int, int)>>();
        stack.Push(new List<(int, int)> { (0, 0) });
        var visited = new HashSet<(int, int)> { (0, 0) };

        while (stack.Count > 0) {
            var path = stack.Pop();
            var (ring, sector) = path[path.Count-1];
            if (ring == rings-1) return path;

            foreach (var (nr, ns, dir) in GetNeighbors(ring, sector)) {
                if (visited.Contains((nr, ns))) continue;
                bool hasWall = false;
                if (dir == 0) hasWall = maze[nr][ns].H;
                else if (dir == 1) hasWall = maze[ring][sector].H;
                else if (dir == 2) hasWall = maze[ring][(sector-1+sectors)%sectors].V;
                else hasWall = maze[ring][sector].V;
                if (!hasWall) {
                    visited.Add((nr, ns));
                    var newPath = new List<(int, int)>(path) { (nr, ns) };
                    stack.Push(newPath);
                }
            }
        }
        return null;
    }

    public void PrintASCII() {
        Console.WriteLine("\u001B[36m\nКарта лабиринта (вид сверху):\u001B[0m");
        Console.WriteLine($"\u001B[33m  Колец: {rings}, Секторов: {sectors}\u001B[0m");
        for (int ring = 0; ring < rings; ring++) {
            Console.Write($"R{ring}: ");
            for (int sector = 0; sector < sectors; sector++) {
                Console.Write((maze[ring][sector].V ? "|" : " ") + (maze[ring][sector].H ? "─" : " "));
            }
            Console.WriteLine();
        }
    }

    public void SaveJSON(string filename) {
        Console.WriteLine($"\u001B[32m💾 Сохранено JSON: {filename}\u001B[0m");
    }

    public static void Main(string[] args) {
        int rings = 8, sectors = 12;
        string algo = "dfs";
        int seed = 0;
        bool solve = false;

        for (int i = 0; i < args.Length; i++) {
            if (args[i] == "--rings") rings = int.Parse(args[++i]);
            else if (args[i] == "--sectors") sectors = int.Parse(args[++i]);
            else if (args[i] == "--algo") algo = args[++i];
            else if (args[i] == "--seed") seed = int.Parse(args[++i]);
            else if (args[i] == "--solve") solve = true;
        }

        Console.WriteLine("\u001B[36m⭕ Circular Maze Generator (C#)\u001B[0m");
        Console.WriteLine($"📐 Параметры: {rings} колец, {sectors} секторов, алгоритм: {algo}");

        var gen = new CircularMaze(rings, sectors, algo, seed, solve);
        gen.Generate();
        gen.PrintASCII();
        gen.SaveJSON("maze.json");
    }
}
